networkx/algorithms/dag.py

```diff
@@ -32,69 +29,32 @@
     "dag_to_branching",
 ]
 
-chaini = chain.from_iterable
-
-
-def descendants(G, source):
-    """Returns all nodes reachable from `source` in `G`.
-
-    Parameters
-    ----------
-    G : NetworkX Graph
-    source : node in `G`
-
-    Returns
-    -------
-    set()
-        The descendants of `source` in `G`
-
-    Raises
-    ------
-    NetworkXError
-        If node `source` is not in `G`.
-
-    Examples
-    --------
-    >>> DG = nx.path_graph(5, create_using=nx.DiGraph)
-    >>> sorted(list(nx.descendants(DG, 2)))
-    [3, 4]
+_deprecated_names = [
+    "descendants",
+    "ancestors",
+    "transitive_closure",
+]
 
-    See also
-    --------
-    ancestors
-    """
-    return {child for parent, child in nx.bfs_edges(G, source)}
 
+def __getattr__(name):
+    f = getattr(nx, name)
 
-def ancestors(G, source):
-    """Returns all nodes having a path to `source` in `G`.
+    if name in _deprecated_names:
+        from warnings import warn
+        from inspect import getmodule
 
-    Parameters
-    ----------
-    G : NetworkX Graph
-    source : node in `G`
+        msg = (
+            f"\nThe function {name} will be moved "
+            f"to {getmodule(f).__name__} in "
+            "networkx 2.7\n"
```

Comment:
I think we'll need to iterate a bit on the warning message to make it more detailed. Some things to include would be recommended access patterns (i.e. "this function is available from the top-level namespace, e.g. `nx.{name}`" or from the reachability module: `from networkx.algorithms.traversal.reachability import {name}`), esp. since users should likely prefer the first.

Also, I think we'd want to switch the version to when this warning will expire. With version 3.0 coming up I think it makes sense to stop warning at that point. So the warning should say something along the lines of "{name} has been moved to {new module}. This move will become permanent in 3.0, ..."

Reply:
Thanks for this - the new warning message is certainly more informative and LGTM, though some other fresh eyes would be great.